pxr/imaging/hdSt/material.cpp

```diff
@@ -172,48 +177,91 @@ HdStMaterial::_ProcessTextureDescriptors(
 /* virtual */
 void
 HdStMaterial::Sync(HdSceneDelegate *sceneDelegate,
-                 HdRenderParam   *renderParam,
-                 HdDirtyBits     *dirtyBits)
+                   HdRenderParam   *renderParam,
+                   HdDirtyBits     *dirtyBits)
 {
     HD_TRACE_FUNCTION();
     HF_MALLOC_TAG_FUNCTION();
 
+    const HdDirtyBits bits = *dirtyBits;
+
+    if (!(bits & DirtyResource) && !(bits & DirtyParams)) {
+        *dirtyBits = Clean;
+        return;
+    }
+
+    bool processedMaterialNetwork = false;
+
     HdStResourceRegistrySharedPtr const& resourceRegistry =
         std::static_pointer_cast<HdStResourceRegistry>(
             sceneDelegate->GetRenderIndex().GetResourceRegistry());
 
-    HdDirtyBits bits = *dirtyBits;
+#ifdef PXR_MATERIALX_SUPPORT_ENABLED
+    {
+        HdStRenderDelegate* stormDelegate = static_cast<HdStRenderDelegate*>(
+            sceneDelegate->GetRenderIndex().GetRenderDelegate());
 
-    if (!(bits & DirtyResource) && !(bits & DirtyParams)) {
-        *dirtyBits = Clean;
-        return;
+        HdSt_MaterialFilterTaskSharedPtr filterTask;
+
+        // This scene index manages early parallel MaterialX codegen. It exists
+        // when that optimization is enabled.
+        if (HdSt_MaterialXSyncSceneIndex* sceneIndex =
+            stormDelegate->GetMaterialXSyncSceneIndex()) {
+
+            // Wait for all early parallel codegen tasks to complete and
+            // retrieve the state cached for this sprim when codegen was
+            // started
+            filterTask = sceneIndex->WaitAndExtractFilterTask(GetId());
```

Comment:
Wait for the codegen tasks to complete. Previously, we were waiting by calling a new render delegate API before syncing the materials: https://github.com/PixarAnimationStudios/OpenUSD/pull/3567/files#r1995705092

Now, we wait for the generator task group in each material's sync. In practice, only the first material sprim would ever wait for any measurable duration, and the overhead of waiting for all subsequent sprims is negligible (around 1 microsecond).

This is also where we get the cached per-material state from the scene index. The filter task is removed from the scene index its ownership is transferred to the local variable.

Reusing the filter task allows us to avoid getting the material resource again below.